### PR TITLE
topology1: Use DYNAMIC for ADL and RPL topologies

### PR DESCRIPTION
**Pull request by Vamshigopal**
use dynamic for all the adl and rpl topologies except non 3p(waves,DTS)

### CODE REVIEW

**Review by lgirdwood — APPROVED** [review]
@ranj063 good for you ?

File: tools/topology/topology1/CMakeLists.txt [hunk 1]
```diff
@@ -61,70 +61,70 @@ set(TPLGS
...
+	"sof-icl-rt711-rt1308-rt715-hdmi\;sof-adl-rt1316-l2-mono-rt714-l3\;-DPLATFORM=adl\;-DMONO\;-DNOJACK\;-DAMP_1_LINK=2\;-DEXT_AMP_REF\;-DMIC_LINK=3\;-DDYNAMIC=1"
+	"sof-icl-rt711-rt1308-rt715-hdmi\;sof-rpl-rt1316-l12-rt714-l0\;-DPLATFORM=rpl\;-DNOJACK\;-DAMP_1_LINK=1\;-DAMP_2_LINK=2\;-DEXT_AMP_REF\;-DMIC_LINK=0\;-DDYNAMIC=1"
 	## end SoundWire NOJACK topologies
 
 	"sof-ehl-rt5660\;sof-ehl-rt5660\;-DHDMI=1"
 	"sof-ehl-rt5660\;sof-ehl-rt5660-nohdmi"
-	"sof-tgl-max98357a-rt5682\;sof-tgl-max98357a-rt5682\;-DCODEC=MAX98357A\;-DFMT=s16le\;-DPLATFORM=tgl\;-DLINUX_MACHINE_DRIVER=sof_rt5682\;-DAMP_SSP=1"
-	"sof-tgl-max98357a-rt5682\;sof-adl-max98357a-rt5682\;-DCODEC=MAX98357A\;-DFMT=s16le\;-DPLATFORM=adl\;-DLINUX_MACHINE_DRIVER=sof_rt5682\;-DAMP_SSP=2"
-	"sof-tgl-max98357a-rt5682\;sof-adl-max98357a-rt5682-rtnr\;-DCODEC=MAX98357A\;-DFMT=s16le\;-DPLATFORM=adl\;-DLINUX_MACHINE_DRIVER=sof_rt5682\;-DAMP_SSP=2\;-DCHANNELS=2\;-DRTNR"
+	"sof-tgl-max98357a-rt5682\;sof-tgl-max98357a-rt5682\;-DCODEC=MAX98357A\;-DFMT=s16le\;-DPLATFORM=tgl\;-DLINUX_MACHINE_DRIVER=sof_rt5682\;-DAMP_SSP=1\;-DDYNAMIC=1"
```
Comment (plbossart) [inline tools/topology/topology1/CMakeLists.txt:83]: this is a TGL topology?

File: tools/topology/topology1/CMakeLists.txt [hunk 2]
```diff
@@ -61,70 +61,70 @@ set(TPLGS
...
-	"sof-tgl-max98357a-rt5682\;sof-adl-max98390-rt5682-google-aec\;-DCODEC=MAX98390\;-DFMT=s32le\;-DPLATFORM=adl\;-DLINUX_MACHINE_DRIVER=sof_rt5682\;-DAMP_SSP=1\;-DBT_OFFLOAD\;-DCHANNELS=2\;-DGOOGLE_RTC_AUDIO_PROCESSING\;-DNOHOTWORD\;-DNO16KDMIC"
-	"sof-tgl-max98357a-rt5682\;sof-adl-max98390-rt5682-rtnr-google-aec\;-DCODEC=MAX98390\;-DFMT=s32le\;-DPLATFORM=adl\;-DLINUX_MACHINE_DRIVER=sof_rt5682\;-DAMP_SSP=1\;-DBT_OFFLOAD\;-DCHANNELS=2\;-DGOOGLE_RTC_AUDIO_PROCESSING\;-DRTNR\;-DNOHOTWORD\;-DNO16KDMIC\;-DSPK_MIC_PERIOD_US=10000"
-	"sof-tgl-max98357a-rt5682\;sof-adl-max98390-ssp2-rt5682-ssp0\;-DCODEC=MAX98390\;-DFMT=s32le\;-DPLATFORM=adl\;-DLINUX_MACHINE_DRIVER=sof_rt5682\;-DAMP_SSP=2"
-	"sof-tgl-max98357a-rt5682\;sof-adl-rt5682\;-DNO_AMP\;-DPLATFORM=adl\;-DLINUX_MACHINE_DRIVER=sof_rt5682"
-	"sof-tgl-max98357a-rt5682\;sof-adl-rt1019-rt5682\;-DCODEC=RT1019\;-DFMT=s24le\;-DPLATFORM=adl\;-DLINUX_MACHINE_DRIVER=sof_rt5682\;-DAMP_SSP=1\;-DSPK_MIC_PERIOD_US=10000\;-DBT_OFFLOAD"
+	"sof-tgl-max98357a-rt5682\;sof-adl-max98390-rt5682\;-DCODEC=MAX98390\;-DFMT=s32le\;-DPLATFORM=adl\;-DLINUX_MACHINE_DRIVER=sof_rt5682\;-DAMP_SSP=1\;-DBT_OFFLOAD\;-DDYNAMIC=1"
```
Comment (plbossart) [inline tools/topology/topology1/CMakeLists.txt:103]: IIRC we don't want the DYNAMIC part when there are volatile registers used by the smart amp.

**Review by plbossart — CHANGES_REQUESTED** [review]
This looks a bit too much of a sledgehammer patch that will break more eggs than intended.

What is the problem we are trying to fix anyways? Why is this necessary and why now?

And there's also a kernel dependency, so there's a risk of breaking existing platforms if they update sof-bin but still use a kernel  < 5.19

File: tools/topology/topology1/CMakeLists.txt (hunk 1 above)
Comment (Vamshigopal) [reply tools/topology/topology1/CMakeLists.txt:83]: ack, it was added by mistake. I'll correct it.

**Comment by Vamshigopal** [comment]
> This looks a bit too much of a sledgehammer patch that will break more eggs than intended.
> 
> What is the problem we are trying to fix anyways? 

With Dynamic pipelines enabled, some of the power KPI show better numbers. We discussed with our customer and have agreed to move to dymanic pipeline  in a phased manner - this is first set for ADL and RPL programs which are already verified.

> Why is this necessary and why now?

We are planning to keep dynamic from cavs2.5 onwards.

> And there's also a kernel dependency, so there's a risk of breaking existing platforms if they update sof-bin but still use a kernel < 5.19

our customer kernels for ADL , ADL- N and RPL programs all have the required kernel support too. Older kernels doesnt have this feature so wont be impacted.

**Comment by Vamshigopal** [comment]
@plbossart  may i know your thoughts please ?

File: tools/topology/topology1/CMakeLists.txt (hunk 2 above)
Comment (johnylin76) [reply tools/topology/topology1/CMakeLists.txt:103]: I think this is only for smart amp (due to volatile register use), other 3p comps should be fine to adopt dynamic pipeline.

Comment (Vamshigopal) [reply tools/topology/topology1/CMakeLists.txt:103]: We have not tested enough on 3p components , so as part1 enabling, we will do it for non smart amp and non 3p pipelines.

Comment (johnylin76) [reply tools/topology/topology1/CMakeLists.txt:103]: I see. Thanks

Comment (johnylin76) [reply tools/topology/topology1/CMakeLists.txt:103]: Maybe we could/should implement the dependency as code. e.g. adding in `tools/topology/topology1/platform/intel/intel-boards.m4`

```
undefine(`HAS_3P_PROC')
ifdef(`WAVES', `define(`HAS_3P_PROC', 1)',
    `ifdef(`DTS', `define(`HAS_3P_PROC', 1)',
        `ifdef(`SMART_AMP', `define(`HAS_3P_PROC', 1)',
            `ifdef(`RTNR', `define(`HAS_3P_PROC', 1)',
                `ifdef(`IGO', `define(`HAS_3P_PROC', 1)',
                    `ifdef(`GOOGLE_RTC_AUDIO_PROCESSING', `define(`HAS_3P_PROC', 1)'
)')')')')')

undefine(`DYNAMIC')
ifelse(PLATFORM, `adl', `ifdef(`HAS_3P_PROC', `', `define(`DYNAMIC', 1)')',
       PLATFORM, `rpl', `ifdef(`HAS_3P_PROC', `', `define(`DYNAMIC', 1)')',
       `')
```

not sure how to flatten it

**Comment by plbossart** [comment]
IIRC there are dependencies on the kernel to support the DYNAMIC pipelines.

I am not sure there is any merit of changing all topologies blindly, clearly you have not tested the Dell/SoundWire topologies so there's at least a very large validation gap....

@ranj063 please chime in.

**Comment by johnylin76** [comment]
> IIRC there are dependencies on the kernel to support the DYNAMIC pipelines.
> 
> I am not sure there is any merit in changing all topologies blindly, clearly you have not tested the Dell/SoundWire topologies so there's at least a very large validation gap....
> 
> @ranj063 please chime in.

I think you made a good point that I just overlooked. Apparently my proposed code lacks many details. Although in my view that is still worth when the new config is targeted to be set massively and can be deduced from the existing ones. 

I don't mind to add the config case by case in CMakeLists.txt since the deduction logic is a bit complicated.

**Review by johnylin76 — APPROVED** [review]
LGTM

**Comment by lgirdwood** [comment]
> I don't mind to add the config case by case in CMakeLists.txt since the deduction logic is a bit complicated.

Can we target this on case by case, IIUC @plbossart can approve that.

**Comment by plbossart** [comment]
> > I don't mind to add the config case by case in CMakeLists.txt since the deduction logic is a bit complicated.
> 
> Can we target this on case by case, IIUC @plbossart can approve that.

I don't mind changes to the Chrome topologies if they have been tested and are compatible with the Chrome kernel.
But don't change all topologies randomly, e..g the ones used by Dell SoundWire devices should not be changed by the Chrome team.

**Comment by Vamshigopal** [comment]
> > > I don't mind to add the config case by case in CMakeLists.txt since the deduction logic is a bit complicated.
> > 
> > 
> > Can we target this on case by case, IIUC @plbossart can approve that.
> 
> I don't mind changes to the Chrome topologies if they have been tested and are compatible with the Chrome kernel. But don't change all topologies randomly, e..g the ones used by Dell SoundWire devices should not be changed by the Chrome team.

Ack, Updated PR excluding Dell SoundWire topologies.

File: tools/topology/topology1/CMakeLists.txt [hunk 3]
```diff
@@ -71,44 +71,44 @@ set(TPLGS
...
-	"sof-icl-rt711-rt1308-rt715-hdmi\;sof-adl-rt1316-l2-mono-rt714-l3\;-DPLATFORM=adl\;-DMONO\;-DNOJACK\;-DAMP_1_LINK=2\;-DEXT_AMP_REF\;-DMIC_LINK=3"
-	"sof-icl-rt711-rt1308-rt715-hdmi\;sof-rpl-rt1316-l12-rt714-l0\;-DPLATFORM=rpl\;-DNOJACK\;-DAMP_1_LINK=1\;-DAMP_2_LINK=2\;-DEXT_AMP_REF\;-DMIC_LINK=0"
+	"sof-icl-rt711-rt1308-rt715-hdmi\;sof-adl-rt1316-l1-mono-rt714-l0\;-DPLATFORM=adl\;-DMONO\;-DNOJACK\;-DAMP_1_LINK=1\;-DEXT_AMP_REF\;-DMIC_LINK=0\;-DDYNAMIC=1"
+	"sof-icl-rt711-rt1308-rt715-hdmi\;sof-adl-rt1316-l2-mono-rt714-l0\;-DPLATFORM=adl\;-DMONO\;-DNOJACK\;-DAMP_1_LINK=2\;-DEXT_AMP_REF\;-DMIC_LINK=0\;-DDYNAMIC=1"
+	"sof-icl-rt711-rt1308-rt715-hdmi\;sof-adl-rt1316-l12-rt714-l0\;-DPLATFORM=adl\;-DNOJACK\;-DAMP_1_LINK=1\;-DAMP_2_LINK=2\;-DEXT_AMP_REF\;-DMIC_LINK=0\;-DDYNAMIC=1"
+	"sof-icl-rt711-rt1308-rt715-hdmi\;sof-adl-rt1316-l2-mono-rt714-l3\;-DPLATFORM=adl\;-DMONO\;-DNOJACK\;-DAMP_1_LINK=2\;-DEXT_AMP_REF\;-DMIC_LINK=3\;-DDYNAMIC=1"
+	"sof-icl-rt711-rt1308-rt715-hdmi\;sof-rpl-rt1316-l12-rt714-l0\;-DPLATFORM=rpl\;-DNOJACK\;-DAMP_1_LINK=1\;-DAMP_2_LINK=2\;-DEXT_AMP_REF\;-DMIC_LINK=0\;-DDYNAMIC=1"
```
Comment (plbossart) [inline tools/topology/topology1/CMakeLists.txt:78]: all of this is for Dell!

**Review by plbossart — CHANGES_REQUESTED** [review]

**Review by plbossart — APPROVED** [review]

**Comment by Vamshigopal** [comment]
Thanks @plbossart for suggestions and reviews.
@lgirdwood can you merge this PR ?

**Comment by kv2019i** [comment]
@Vamshigopal I'll proceed with merge. Can you make a backport to stable-v2.2 ? I will soon remove this file from mainline as we no longer support IPC3 for Intel targets in mainline, so any patch that is not in stable-v2.2 is at risk of getting lost.

**Comment by Vamshigopal** [comment]
> @Vamshigopal I'll proceed with merge. Can you make a backport to stable-v2.2 ? I will soon remove this file from mainline as we no longer support IPC3 for Intel targets in mainline, so any patch that is not in stable-v2.2 is at risk of getting lost.

ack, #8229 for stable-v2.2 branch

**Pull request merged**